.github/workflows/build_and_test_dbt.yaml

```diff
@@ -0,0 +1,122 @@
+
+      - name: Load environment variables
+        uses: ./.github/actions/load_environment_variables
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v2
+        with:
+          role-to-assume: ${{ secrets.AWS_IAM_ROLE_TO_ASSUME_ARN }}
+          aws-region: us-east-1
+
+      - name: Set environment for branch
+        run: |
+          if [[ $GITHUB_REF_NAME == 'master' ]]; then
+            echo "On master branch"
+            {
+              echo "TARGET=prod";
+              echo "CACHE_KEY=master";
+            } >> "$GITHUB_ENV"
+          elif [[ $GITHUB_REF_NAME == 'data-catalog' ]]; then
+            echo "On data catalog branch"
+            {
+              echo "TARGET=ci";
+              echo "CACHE_KEY=data-catalog";
+              echo "GITHUB_HEAD_REF=data-catalog";
+            } >> "$GITHUB_ENV"
+          else
+            echo "On pull request branch"
+            {
+              echo "TARGET=ci";
+              echo "CACHE_KEY=$GITHUB_HEAD_REF";
+            } >> "$GITHUB_ENV"
+          fi
+        shell: bash
+
+      - name: Cache dbt manifest
+        id: cache
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.MANIFEST_DIR }}
+          key: ${{ env.CACHE_NAME }}-${{ env.CACHE_KEY }}
+          restore-keys: |
+            ${{ env.CACHE_NAME }}-data-catalog
+            ${{ env.CACHE_NAME }}-master
+
+      - if: ${{ steps.cache.outputs.cache-hit == 'true' }}
+        name: Set command args to build/test modified resources
+        run: echo "MODIFIED_RESOURCES_ONLY=true" >> "$GITHUB_ENV"
+        shell: bash
+
+      - if: ${{ steps.cache.outputs.cache-hit != 'true' }}
+        name: Set command args to build/test all resources
+        run: echo "MODIFIED_RESOURCES_ONLY=false" >> "$GITHUB_ENV"
+        shell: bash
+
+      - name: Test dbt macros
+        run: dbt run-operation test_all
+        working-directory: ${{ env.PROJECT_DIR }}
+        shell: bash
+
+      - name: Build models
+        run: |
+          if [[ $MODIFIED_RESOURCES_ONLY == 'true' ]]; then
+            echo "Running build on modified resources only"
+            dbt run --target "$TARGET" -s state:modified --defer --state target/
+          else
+            echo "Running build on all resources"
+            dbt run --target "$TARGET"
+          fi
+        working-directory: ${{ env.PROJECT_DIR }}
+        shell: bash
+
+      - name: Test models
+        run: |
+          if [[ $MODIFIED_RESOURCES_ONLY == 'true' ]]; then
+            echo "Running tests on modified resources only"
+            dbt test --target "$TARGET" -s state:modified --state target/
+          else
+            echo "Running tests on all resources"
+            dbt test --target "$TARGET"
+          fi
+        working-directory: ${{ env.PROJECT_DIR }}
+        shell: bash
+
+      - if: ${{ github.event_name == 'pull_request' && github.event.action == 'closed' }}
```

Comment:
**suggestion (blocking):** Maybe factor these out into a separate workflow if possible. Currently closing/merging a PR would re-run the entire build and test process, which will happen anyways in the prod env when merged to master.

There's also a risk that someone does something silly and gets their branch into a non-zero exit state in the builds/tests, see https://github.com/ccao-data/data-architecture/pull/52. If they then wanted to abandon/close that branch without merging and delete the ci env, they would need to reconcile their branch until everything runs in order to reach this step.

Reply:
Smart, done in 704ed13d79918028c3bc0e66daec37ac759be9fc!

Reply:
Here's a test run demonstrating that the factored-out workflow still works: https://github.com/ccao-data/data-architecture/actions/runs/5799412187/job/15719353357